File: experiments/adrianco/experiment-13/runs/language=go_model=opus-4.8-fast_prompt=TDD/rep1/mcp_test.go
// Package main — Brazilian Soccer MCP Server.
//
// mcp_test.go: Tests the MCP JSON-RPC layer — initialize handshake, tools/list
// discovery, tools/call dispatch for every tool, notification handling and
// error responses — plus the human-readable formatting of each tool result.
package main

import (
	"encoding/json"
	"strings"
	"testing"
)

func newServer(t *testing.T) *Server {
	t.Helper()
	return NewServer(testDataset(t))
}

// call is a helper that sends one JSON-RPC request and returns the decoded
// response object (nil if the server produced no response, e.g. notifications).
func call(t *testing.T, s *Server, req string) map[string]any {
	t.Helper()
	out, ok := s.Dispatch([]byte(req))
	if !ok {
		return nil
	}
	var resp map[string]any
	if err := json.Unmarshal(out, &resp); err != nil {
		t.Fatalf("invalid JSON response %q: %v", out, err)
	}
	return resp
}

func TestInitialize(t *testing.T) {
	s := newServer(t)
	resp := call(t, s, `{"jsonrpc":"2.0","id":1,"method":"initialize","params":{}}`)
	if resp["jsonrpc"] != "2.0" {
		t.Errorf("missing jsonrpc version: %v", resp)
	}
	result, ok := resp["result"].(map[string]any)
	if !ok {
		t.Fatalf("no result object: %v", resp)
	}
	if _, ok := result["protocolVersion"]; !ok {
		t.Error("initialize result missing protocolVersion")
	}
	info, ok := result["serverInfo"].(map[string]any)
	if !ok || info["name"] == "" {
		t.Errorf("missing serverInfo.name: %v", result)
	}
}

func TestNotificationHasNoResponse(t *testing.T) {
	s := newServer(t)
	out, ok := s.Dispatch([]byte(`{"jsonrpc":"2.0","method":"notifications/initialized"}`))
	if ok {
		t.Errorf("notification should produce no response, got %s", out)
	}
}

func TestToolsList(t *testing.T) {
	s := newServer(t)
	resp := call(t, s, `{"jsonrpc":"2.0","id":2,"method":"tools/list","params":{}}`)
	result := resp["result"].(map[string]any)
	tools, ok := result["tools"].([]any)
	if !ok || len(tools) < 6 {
		t.Fatalf("expected >= 6 tools, got %v", result["tools"])
	}
	// Every tool must have name, description and an object inputSchema.
	names := map[string]bool{}
	for _, raw := range tools {
		tool := raw.(map[string]any)
		name, _ := tool["name"].(string)
		if name == "" {
			t.Error("tool with empty name")
		}
		names[name] = true
		if _, ok := tool["inputSchema"].(map[string]any); !ok {
			t.Errorf("tool %q missing inputSchema object", name)
		}
		if tool["description"] == "" {
			t.Errorf("tool %q missing description", name)
		}
	}
	for _, want := range []string{"search_matches", "team_record", "head_to_head", "standings", "search_players", "competition_stats"} {
		if !names[want] {
			t.Errorf("missing expected tool %q", want)
		}
	}
}

// callTool invokes a tool and returns its concatenated text content.
func callTool(t *testing.T, s *Server, name string, args map[string]any) (string, bool) {
	t.Helper()
	params := map[string]any{"name": name, "arguments": args}
	reqObj := map[string]any{"jsonrpc": "2.0", "id": 9, "method": "tools/call", "params": params}
	raw, _ := json.Marshal(reqObj)
	out, ok := s.Dispatch(raw)
	if !ok {
		t.Fatal("tools/call produced no response")
	}
	var resp map[string]any
	if err := json.Unmarshal(out, &resp); err != nil {
		t.Fatalf("bad response: %v", err)
	}
	if errObj, ok := resp["error"].(map[string]any); ok {
		t.Fatalf("unexpected JSON-RPC error: %v", errObj)
	}
	result := resp["result"].(map[string]any)
	isErr, _ := result["isError"].(bool)
	content := result["content"].([]any)
	var sb strings.Builder
	for _, c := range content {
		cm := c.(map[string]any)
		if cm["type"] == "text" {
			sb.WriteString(cm["text"].(string))
		}
	}
	return sb.String(), isErr
}

func TestCallSearchMatches(t *testing.T) {
	s := newServer(t)
	text, isErr := callTool(t, s, "search_matches", map[string]any{
		"team": "Flamengo", "opponent": "Fluminense",
	})
	if isErr {
		t.Fatalf("unexpected error result: %s", text)
	}
	if !strings.Contains(strings.ToLower(text), "flamengo") || !strings.Contains(strings.ToLower(text), "fluminense") {
		t.Errorf("match output missing teams: %s", text)
	}
}

func TestCallStandings(t *testing.T) {
	s := newServer(t)
	text, isErr := callTool(t, s, "standings", map[string]any{
		"competition": "Brasileirão", "season": float64(2019),
	})
	if isErr {
		t.Fatalf("unexpected error: %s", text)
	}
	if !strings.Contains(strings.ToLower(text), "flamengo") || !strings.Contains(text, "90") {
		t.Errorf("standings output wrong: %s", text)
	}
}

func TestCallTeamRecord(t *testing.T) {
	s := newServer(t)
	text, _ := callTool(t, s, "team_record", map[string]any{
		"team": "Flamengo", "season": float64(2019), "source": "Brasileirao",
	})
	if !strings.Contains(text, "28") {
		t.Errorf("team record missing 28 wins: %s", text)
	}
}

func TestCallSearchPlayers(t *testing.T) {
	s := newServer(t)
	text, _ := callTool(t, s, "search_players", map[string]any{
		"nationality": "Brazil", "limit": float64(3),
	})
	if !strings.Contains(text, "Neymar") {
		t.Errorf("expected Neymar among top Brazilians: %s", text)
	}
}

func TestCallCompetitionStats(t *testing.T) {
	s := newServer(t)
	text, _ := callTool(t, s, "competition_stats", map[string]any{
		"source": "Brasileirao", "season": float64(2019),
	})
	low := strings.ToLower(text)
	if !strings.Contains(low, "average") && !strings.Contains(low, "avg") {
		t.Errorf("stats output missing average goals: %s", text)
	}
}

func TestCallHeadToHead(t *testing.T) {
	s := newServer(t)
	text, _ := callTool(t, s, "head_to_head", map[string]any{
		"team_a": "Palmeiras", "team_b": "Santos",
	})
	low := strings.ToLower(text)
	if !strings.Contains(low, "palmeiras") || !strings.Contains(low, "santos") {
		t.Errorf("h2h missing teams: %s", text)
	}
}

func TestCallUnknownToolIsError(t *testing.T) {
	s := newServer(t)
	text, isErr := callTool(t, s, "no_such_tool", map[string]any{})
	if !isErr {
		t.Errorf("expected isError for unknown tool, got: %s", text)
	}
}

func TestMissingRequiredArgIsError(t *testing.T) {
	s := newServer(t)
	// team_record requires a team name.
	text, isErr := callTool(t, s, "team_record", map[string]any{})
	if !isErr {
		t.Errorf("expected error for missing team arg, got: %s", text)
	}
}

func TestUnknownMethodReturnsError(t *testing.T) {
	s := newServer(t)
	resp := call(t, s, `{"jsonrpc":"2.0","id":7,"method":"does/not/exist"}`)
	if _, ok := resp["error"].(map[string]any); !ok {
		t.Errorf("expected JSON-RPC error for unknown method: %v", resp)
	}
}

File: experiments/adrianco/experiment-4/runs/language=clojure_model=claude-opus-4-8_tooling=none/rep2/test/brazilian_soccer/mcp_test.clj
(ns brazilian-soccer.mcp-test
  "Context
  =======
  BDD tests for the MCP/JSON-RPC layer: protocol handshake, tool discovery and
  tool invocation, exercising `handle-request` directly (transport-independent)."
  (:require [clojure.test :refer [deftest testing is]]
            [clojure.string :as str]
            [brazilian-soccer.mcp :as mcp]))

(deftest initialize-handshake
  (testing "Scenario: client initialises the server"
    ;; When the client sends initialize
    (let [resp (mcp/handle-request {:jsonrpc "2.0" :id 1 :method "initialize"})]
      ;; Then the server returns protocol version, capabilities and info
      (is (= 1 (:id resp)))
      (is (= mcp/protocol-version (get-in resp [:result :protocolVersion])))
      (is (contains? (get-in resp [:result :capabilities]) :tools))
      (is (= "brazilian-soccer-mcp" (get-in resp [:result :serverInfo :name]))))))

(deftest notifications-have-no-response
  (testing "Scenario: notifications/initialized produces no reply"
    (is (nil? (mcp/handle-request {:jsonrpc "2.0" :method "notifications/initialized"})))))

(deftest tools-are-listed
  (testing "Scenario: tools/list advertises the tool catalogue"
    (let [resp (mcp/handle-request {:jsonrpc "2.0" :id 2 :method "tools/list"})
          names (set (map :name (get-in resp [:result :tools])))]
      (is (contains? names "find_matches"))
      (is (contains? names "team_record"))
      (is (contains? names "standings"))
      (is (contains? names "search_players"))
      ;; every tool has a name, description and input schema
      (doseq [t (get-in resp [:result :tools])]
        (is (string? (:name t)))
        (is (string? (:description t)))
        (is (= "object" (get-in t [:inputSchema :type])))))))

(defn- call [tool args]
  (mcp/handle-request {:jsonrpc "2.0" :id 9 :method "tools/call"
                       :params {:name tool :arguments args}}))

(defn- text [resp] (get-in resp [:result :content 0 :text]))

(deftest call-find-matches
  (testing "Scenario: tools/call find_matches returns formatted text"
    (let [resp (call "find_matches" {:team "Flamengo" :opponent "Fluminense"})]
      (is (false? (get-in resp [:result :isError])))
      (is (string? (text resp)))
      (is (str/includes? (text resp) "Flamengo")))))

(deftest call-standings
  (testing "Scenario: tools/call standings renders the league table"
    (let [resp (call "standings" {:competition "Brasileirão Série A" :season 2019})]
      (is (false? (get-in resp [:result :isError])))
      (is (str/includes? (text resp) "Flamengo"))
      (is (str/includes? (text resp) "90 pts")))))

(deftest call-team-record
  (testing "Scenario: tools/call team_record renders the record block"
    (let [resp (call "team_record" {:team "Flamengo" :season 2019
                                     :competition "Brasileirão Série A"})]
      (is (str/includes? (text resp) "Wins: 28"))
      (is (str/includes? (text resp) "Win rate: 73.7%")))))

(deftest call-search-players
  (testing "Scenario: tools/call search_players renders a ranked list"
    (let [resp (call "search_players" {:nationality "Brazil" :limit 3})]
      (is (str/includes? (text resp) "Neymar")))))

(deftest unknown-tool-is-an-error-result
  (testing "Scenario: calling an unknown tool yields an isError result"
    (let [resp (call "does_not_exist" {})]
      (is (true? (get-in resp [:result :isError])))
      (is (str/includes? (text resp) "Unknown tool")))))

(deftest unknown-method-returns-jsonrpc-error
  (testing "Scenario: an unknown method returns a JSON-RPC error object"
    (let [resp (mcp/handle-request {:jsonrpc "2.0" :id 7 :method "no/such"})]
      (is (= -32601 (get-in resp [:error :code]))))))
